src/puya/resolve-puya-path.ts

```diff
@@ -1,17 +1,32 @@
+import * as fs from 'fs'
 import { Constants } from '../constants'
 import { logger } from '../logger'
-import { downloadPuyaBinary, findCachedPuyaBinary } from './puya-binary'
+import { createLockFile } from '../util/lock-file'
+import { downloadPuyaBinary, getCachedPuyaBinaryPath, getPuyaStorageDir } from './puya-binary'
 import { parseSemVer } from './semver'
 
 export async function resolvePuyaPath(): Promise<string> {
   const version = parseSemVer(Constants.targetedPuyaVersion)
 
-  const cachedBinaryPath = findCachedPuyaBinary(version)
-  if (cachedBinaryPath) {
-    logger.debug(undefined, `Found cached Puya binary at ${cachedBinaryPath}`)
+  const puyaStorageDir = getPuyaStorageDir()
+  const cachedBinaryPath = getCachedPuyaBinaryPath(puyaStorageDir, version)
+
+  if (checkIfCacheExists(cachedBinaryPath)) {
     return cachedBinaryPath
   }
 
-  const downloadedPath = await downloadPuyaBinary(version)
-  return downloadedPath
+  await using _ = await createLockFile(`${puyaStorageDir}.lock`, { maxRetries: 30, delayMs: 1000, staleMs: 60 * 1000 })
+
+  if (checkIfCacheExists(cachedBinaryPath)) {
```

Comment:
Probably worth a small comment here. Something like:
"Between the first check and acquiring the lock, the binary may have been downloaded by another process. In this case there is no need to download the binary again."